core/gallery/service.go

```diff
@@ -0,0 +1,494 @@
+	archiveDownloadingPercents = 30
+	archiveCopyingPercents     = 10
+
+	indexName = "app-index.json"
+)
+
+type builtInUseCaseInfo struct {
+	Title, DownloadLink string
+}
+
+var (
+	log = logger.NewNamed(CName)
+
+	// TODO: GO-4131 Fill in download links when built-in usecases will be downloaded to gallery
+	ucCodeToInfo = map[pb.RpcObjectImportUseCaseRequestUseCase]builtInUseCaseInfo{
+		pb.RpcObjectImportUseCaseRequest_GET_STARTED:       {"Get Started", "https://storage.gallery.any.coop/get_started/get_started.zip"},
+		pb.RpcObjectImportUseCaseRequest_PERSONAL_PROJECTS: {"Personal Projects", ""},
+		pb.RpcObjectImportUseCaseRequest_KNOWLEDGE_BASE:    {"Knowledge Base", ""},
+		pb.RpcObjectImportUseCaseRequest_NOTES_DIARY:       {"Notes and Diary", ""},
+		pb.RpcObjectImportUseCaseRequest_STRATEGIC_WRITING: {"Strategic Writing", ""},
+		pb.RpcObjectImportUseCaseRequest_EMPTY:             {"Empty", ""},
+	}
+
+	errOutdatedArchive = fmt.Errorf("archive is outdated")
+)
+
+type Service interface {
+	app.Component
+
+	ImportExperience(ctx context.Context, spaceID, url, title, cachePath string, newSpace bool) (err error)
+	ImportBuiltInUseCase(
+		ctx context.Context, spaceID, cachePath string, req pb.RpcObjectImportUseCaseRequestUseCase,
+	) (code pb.RpcObjectImportUseCaseResponseErrorCode, err error)
+
+	GetGalleryIndex(clientCachePath string) (*pb.RpcGalleryDownloadIndexResponse, error)
+	GetManifest(url string, checkWhitelist bool) (info *model.ManifestInfo, err error)
+}
+
+type service struct {
+	importer        importer.Importer
+	spaceNameGetter objectstore.SpaceNameGetter
+	tempDirService  core.TempDirProvider
+	progress        process.Service
+	notifications   notifications.Notifications
+	indexCache      IndexCache
+
+	withUrlValidation bool
+}
+
+func New() Service {
+	return &service{}
+}
+
+func (s *service) Init(a *app.App) (err error) {
+	s.importer = a.MustComponent(importer.CName).(importer.Importer)
+	s.spaceNameGetter = app.MustComponent[objectstore.SpaceNameGetter](a)
+	s.tempDirService = app.MustComponent[core.TempDirProvider](a)
+	s.progress = a.MustComponent(process.CName).(process.Service)
+	s.notifications = app.MustComponent[notifications.Notifications](a)
+	s.indexCache = app.MustComponent[IndexCache](a)
+
+	s.withUrlValidation = true
+	return
+}
+
+func (s *service) Name() string {
+	return CName
+}
+
+func (s *service) ImportBuiltInUseCase(
+	ctx context.Context,
+	spaceID, cachePath string,
+	useCase pb.RpcObjectImportUseCaseRequestUseCase,
+) (code pb.RpcObjectImportUseCaseResponseErrorCode, err error) {
+	if useCase == pb.RpcObjectImportUseCaseRequest_NONE {
+		return pb.RpcObjectImportUseCaseResponseError_NULL, nil
+	}
+
+	start := time.Now()
+
+	info, found := ucCodeToInfo[useCase]
+	if !found {
+		return pb.RpcObjectImportUseCaseResponseError_BAD_INPUT,
+			fmt.Errorf("failed to import built-in usecase: invalid Use Case value: %v", useCase)
+	}
+
+	if cachePath == "" {
+		if ctx == nil {
```

Comment:
May be if `ImportBuiltInUseCase` should be called with context.Background() instead of nil to avoid such check?

Reply:
Yeah, I have added ctx to all calls, thank you!